### PR TITLE
Feature/6549 add vaccination tile on home screen

### PR DESCRIPTION
**Pull request by 30mar**
## Description

- Add Vaccination Tile on the HomeScreen
-  Add Vaccination registration tile on the HomeScreen

TBD:

- Gradient color for the Vaccination tile
- Unit and UITests

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6549

## Screenshots

https://user-images.githubusercontent.com/15270737/117811771-425e9a00-b261-11eb-952e-651522f73261.MOV

### CODE REVIEW

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 1]
```diff
@@ -2467,3 +2467,15 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"VaccinationCertificates_Home_Title" = "Digitaler Impfnachweis";
+
+"VaccinationCertificates_Home_Body" = "SARS-CoV-2\nImpfschutz";
+
+"VaccinationCertificates_Home_InProgress" = "Unvollständiger\nImpfschutz";
+
+"HealthCertificate_Info_registerCard_title" = "Impfzertifikat registrieren";
+
+"HealthCertificate_Info_registerCard_description" = "Registrieren Sie Impfzertifikate um einen Impfnachweis zu ehalten. Scannen Sie dafür die QR Codes Ihres Dokuments.";
```
Comment (Ein-Tim) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2479]: QR code with a hyphen here to be consistent with event-registration feature?

```suggestion
"HealthCertificate_Info_registerCard_description" = "Registrieren Sie Impfzertifikate um einen Impfnachweis zu ehalten. Scannen Sie dafür die QR-Codes Ihres Dokuments.";
```

Comment (30mar) [reply src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2479]: reference: https://www.figma.com/file/KuDEcMSREg3VhZRhC3FXTY/COVID19_Master_2.0?node-id=13360%3A5470

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

**Comment by KaiTeuber** [comment]
Unittest are missing for the ViewModels

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			self.isVerified = isValid
+			onUpdate()
+		}.store(in: &subscriptions)
+		DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
+			self.isVerified = true
+			onUpdate()
+		}
+		healthCertifiedPerson.$proofCertificate.sink(receiveValue: { proofCertificate in
+			guard let proofCertificate = proofCertificate else {
+				self.vaccinatedPersonName = healthCertifiedPerson.healthCertificates.first?.name.fullName
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift:27]: Would be better to check for existence before assigning.

**Comment by naveeddotio** [comment]
Couldn't see any UITest, is that a part of another ticket? Also, Unit tests are also missing.

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift (hunk 2 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift:27]: we have to use the value  if there was no proof certificate available

**Comment by 30mar** [comment]
@KaiTeuber  @naveeddotio  I mentioned in the PR description there will be a separate branch for the testing and also for a small UI adjustment regarding gradiants

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	init(
+		healthCertifiedPerson: HealthCertifiedPerson,
+		onUpdate: @escaping () -> Void
+	) {
+		self.healthCertifiedPerson = healthCertifiedPerson
+		healthCertifiedPerson.$hasValidProofCertificate.sink { isValid in
+			self.isVerified = isValid
+			onUpdate()
+		}.store(in: &subscriptions)
+		DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
```
[the remark this answers is not in the document]
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift]: please also change it to `.receive(on: DispatchQueue.main.ocombine)`

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift [hunk 4]
```diff
@@ -384,6 +386,14 @@ class HomeTableViewController: UITableViewController, NavigationBarOpacityDelega
 			UINib(nibName: String(describing: HomeRiskTableViewCell.self), bundle: nil),
 			forCellReuseIdentifier: String(describing: HomeRiskTableViewCell.self)
 		)
+		tableView.register(
+			UINib(nibName: String(describing: HomeVaccinationTableViewCell.self), bundle: nil),
+			forCellReuseIdentifier: String(describing: HomeVaccinationTableViewCell.self)
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift:391]: we have a protocol for the reuseIdentifier ˚ReuseIdentifierProviding˚, if the cell fulfills it you can use .reuseIdentifier

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift [hunk 5]
```diff
@@ -56,6 +56,14 @@ class HomeTableViewController: UITableViewController, NavigationBarOpacityDelega
 				self?.viewModel.isUpdating = false
 			}
 			.store(in: &subscriptions)
+		
+		viewModel.$healthCertifiedPersons
+			.receive(on: DispatchQueue.OCombine(.main))
+			.sink { [weak self] _ in
+				self?.tableView.reloadSections([HomeTableViewModel.Section.healthCertificate.rawValue], with: .none)
+				self?.viewModel.isUpdating = false
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift:64]: Does this have any effect?  Where is it set to true for the case of healthCertifiedPersons?

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+			onUpdate()
+		}.store(in: &subscriptions)
+		healthCertifiedPerson.$proofCertificate.sink(receiveValue: { proofCertificate in
+			guard let proofCertificate = proofCertificate else {
+				self.vaccinatedPersonName = healthCertifiedPerson.healthCertificates.first?.name.fullName
+				onUpdate()
+				return
+			}
+			self.vaccinatedPersonName = proofCertificate.name.fullName
+			onUpdate()
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift:28]:
```suggestion
			self.vaccinatedPersonName = healthCertifiedPerson.fullName
			onUpdate()
```

I recently introduced a `fullName` property on the healthCertified person, you can also directly use that one.

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		super.setHighlighted(highlighted, animated: animated)
+
+		containerView.setHighlighted(highlighted, animated: animated)
+	}
+	// MARK: - Internal
+
+	func configure(with cellModel: HomeVaccinationCellModel) {
+		guard !isConfigured else { return }
+
+		cellModel.$vaccinatedPersonName.assign(to: \.text, on: nameLabel).store(in: &subscriptions)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationTableViewCell.swift:30]:
```suggestion
		cellModel.$vaccinatedPersonName
			.receive(on: DispatchQueue.main.ocombine)
			.assign(to: \.text, on: nameLabel)
			.store(in: &subscriptions)
```

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationTableViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+					self?.containerView.backgroundColor = .enaColor(for: .riskNeutral)
+					self?.inProgressLabel.isHidden = false
+					self?.iconView.image = UIImage(named: "Vacc_Incomplete")
+				}
+			}
+			
+		}.store(in: &subscriptions)
+		
+		isConfigured = true
+	}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationTableViewCell.swift:48]:
```suggestion
		cellModel.$isVerified
			.receive(on: DispatchQueue.main.ocombine)
			.sink { [weak self] isVerified in
				if isVerified {
					self?.containerView.backgroundColor = .enaColor(for: .buttonPrimary)
					self?.inProgressLabel.isHidden = true
					self?.iconView.image = UIImage(named: "Vaccination_full")
				} else {
					self?.containerView.backgroundColor = .enaColor(for: .riskNeutral)
					self?.inProgressLabel.isHidden = false
					self?.iconView.image = UIImage(named: "Vacc_Incomplete")
				}
			}
			.store(in: &subscriptions)
		
		isConfigured = true
	}
```

You can directly receive on the main queue.
Additional suggestion: make the three properties publishers on the view model and assign them all separately here.

**Review by nickguendling — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift (hunk 6 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Vaccination/HomeVaccinationCellModel.swift:28]: so i don't have to check if a proof certificate exists ?

**Review by KaiTeuber — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift [hunk 9]
```diff
@@ -56,6 +56,14 @@ class HomeTableViewController: UITableViewController, NavigationBarOpacityDelega
 				self?.viewModel.isUpdating = false
 			}
 			.store(in: &subscriptions)
+		
+		viewModel.$healthCertifiedPersons
+			.receive(on: DispatchQueue.OCombine(.main))
+			.sink { [weak self] _ in
+				self?.tableView.reloadSections([HomeTableViewModel.Section.healthCertificate.rawValue], with: .none)
```
[the remark this answers is not in the document]
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift:63]: But why no animation? `.fade` or `.automatic` are just fine.

File: src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift (hunk 4 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Scenes/Home/HomeTableViewController.swift]: Is this new?

**Pull request merged**